cdc/processor/sinkmanager/manager_impl.go

```diff
@@ -0,0 +1,183 @@
+	// if redo log is enabled.
+	redoManager redo.LogManager
+	// lastBarrierTs is the last barrier ts.
+	lastBarrierTs atomic.Uint64
+	// taskTicker is used to generate fetch tasks.
+	taskTicker *time.Ticker
+	// workers used to pull data from source manager.
+	workers []worker
+	// tableTaskChan is used to send tasks to workers.
+	tableTaskChan chan *tableSinkTask
+	// closedChan is used to notify the manager is closed.
+	closedChan chan struct{}
+}
+
+// New creates a new sink manager.
+func New(redoManager redo.LogManager) Manager {
+	m := &ManagerImpl{
+		progressHeap:  newTableFetchProgress(),
+		tableSinks:    sync.Map{},
+		redoManager:   redoManager,
+		tableTaskChan: make(chan *tableSinkTask),
+		taskTicker:    time.NewTicker(defaultGenerateTaskInterval),
+	}
+
+	// TODO create workers and start them.
+
+	return m
+}
+
+// generateTableSinkFetchTask generates a fetch tableSinkTask to fetch data from the source manager.
+func (m *ManagerImpl) generateTableSinkFetchTask() error {
+	for {
+		select {
+		case <-m.closedChan:
+			return nil
+		case <-m.taskTicker.C:
+			slowestTableProgress := m.progressHeap.pop()
+			tableSink, ok := m.tableSinks.Load(slowestTableProgress.tableID)
+			if !ok {
+				// TODO log it and continue.
+				// Maybe the table sink is removed by the processor.
+				// So we do **not** need add it back to the heap.
+				continue
+			}
+			// If redo log is disabled, we use the barrier ts as the upper bound of the fetch tableSinkTask.
+			// Because can not exceed the barrier ts.
+			barrierTs := m.lastBarrierTs.Load()
+			upperBoundPos := sorter.Position{
+				StartTs:  barrierTs - 1,
+				CommitTs: barrierTs,
+			}
+			canBeAdvance := slowestTableProgress.nextLowerBoundPos.Less(upperBoundPos)
+			if !canBeAdvance || !m.memQuota.TryAcquire() {
+				m.progressHeap.push(slowestTableProgress)
+				continue
+			}
+			tableID := slowestTableProgress.tableID
+			callback := func(lastWrittenPos sorter.Position) {
+				p := progress{
+					tableID:           tableID,
+					nextLowerBoundPos: lastWrittenPos.Next(),
+				}
+				m.progressHeap.push(p)
+			}
+
+			t := &tableSinkTask{
+				tableID:          tableID,
+				lowerBound:       slowestTableProgress.nextLowerBoundPos,
+				upperBound:       upperBoundPos,
+				currentBarrierTs: m.lastBarrierTs.Load(),
+				tableSink:        tableSink.(*tableSinkWrapper),
+				callback:         callback,
+			}
+			select {
+			case <-m.closedChan:
+				return nil
+			case m.tableTaskChan <- t:
+			}
+		}
+	}
+}
+
+// UpdateTableResolvedTs updates the resolved ts of the table.
+func (m *ManagerImpl) UpdateTableResolvedTs(tableID model.TableID, ts model.Ts) {
+	tableSink, ok := m.tableSinks.Load(tableID)
+	if !ok {
+		panic("table sink not found")
+	}
+	tableSink.(*tableSinkWrapper).updateCurrentSorterResolvedTs(ts)
+}
+
+// UpdateBarrierTs updates the barrier ts of all tables in the sink manager.
+func (m *ManagerImpl) UpdateBarrierTs(ts model.Ts) {
+	if ts <= m.lastBarrierTs.Load() {
```

Comment:
If the check is really necessary, we should use
```go
for {
    old := lastBarrierTs.Load()
    if old < new && lastBarrierTs.CompareAndSwap(old, new) {
        break
    }
}
```

Reply:
It seems we only use this method in the processor, so maybe this is acceptable. Other threads only read it.

Reply:
If so it's better to comment it on the function.

Reply:
Added. Please check out https://github.com/pingcap/tiflow/pull/7388/commits/87ed02fbf93089d1dfa8101c5f0a097a63b48f18